File: graph/1319-makeConnected.py
def makeConnected(n,connections):
    if len(connections)< n-1:
        return -1
    
    #create graph now
    g=[[] for _ in range(n)]  # the g=[[]] can not work
    
    for c in connections:
        g[c[0]].append(c[1])
        g[c[1]].append(c[0])
    
    seen=set()
    def dfs(cur):
        seen.add(cur)
        for nxt in g[cur]:
            if nxt not in seen:
                seen.add(nxt)
                dfs(nxt)
    count=0
    for i in range(n):
        if i not in seen:
            count+=1
            dfs(i)
    return count-1

n = 4
connections = [[0,1],[0,2],[1,2]]
print(makeConnected(n,connections))

File: graph/1319-makeConnected.cpp
/*
There are n computers numbered from 0 to n-1 connected by ethernet cables connections forming a network where connections[i] = [a, b] represents a connection between computers a and b. Any computer can reach any other computer directly or indirectly through the network.

Given an initial computer network connections. You can extract certain cables between two directly connected computers, and place them between any pair of disconnected computers to make them directly connected. Return the minimum number of times you need to do this in order to make all the computers connected. If it’s not possible, return -1. 

Example 1:
Input: n = 4, connections = [[0,1],[0,2],[1,2]]
Output: 1
Explanation: Remove cable between computer 1 and 2 and place between computers 1 and 3.
*/
#include <functional>
#include <iostream>
#include <vector>


using namespace std;

class Solution {
public:
  int makeConnected(int n, vector<vector<int>>& connections) {
    if (int(connections.size()) < n - 1)
      return -1;
    vector<vector<int>> g(n);
    for (const auto& c : connections) {  //create the un-directed graph
      g[c[0]].push_back(c[1]);
      g[c[1]].push_back(c[0]);
    }
    vector<int> seen(n);
    int count = 0;
    function<void(int)> dfs = [&](int cur) {
      for (int nxt : g[cur])
        if (!seen[nxt]++) dfs(nxt);      
    };
    for (int i = 0; i < n; ++i)
      if (!seen[i]++ && ++count)
        dfs(i);        
    return count - 1;
  }
};

int main() {
  int n = 4;
  vector<vector<int>> connections{{0, 1}, {0, 2}, {1, 2}};
  cout << Solution().makeConnected(n, connections);
}